File: クラスカル法.cpp
// https://www.slideshare.net/iwiwi/ss-3578491

#include <bits/stdc++.h>
using namespace std;

int N = 3;
int E = 3; // 経路の数
int par[3];
vector<vector<int>> edges = {{20,0,1},{3,1,2},{100,2,0}};

int find(int x) {
	if (par[x] == x) return x;
	return par[x] = find(par[x]);
}

bool same(int x, int y) {
	return find(x) == find(y);
}

void uni(int x, int y) {
	x = find(x);
	y = find(y);
	if (x == y) return;
	par[x] = y;
}

int kruskal() {
	int res = 0;
	for (int i = 0; i < N; ++i) {
		if (!same(edges[i][1],edges[i][2])) {
			res += edges[i][0];
			uni(edges[i][1],edges[i][2]);
		}
	}
	return res;
}

int main() {
	sort(edges.begin(), edges.end());
	for (int i = 0; i < N; ++i) par[i] = i;
	cout << kruskal() << endl;

	return 0;
}

File: Ruby/クラスカル法.rb
# https://www.slideshare.net/iwiwi/ss-3578491

N = 3
par = Array.new(N)

N.times do |n|
	par[n] = n
end

E = 3 # 経路の数

def find(x,par)
	return x if par[x] == x
	return par[x] = find(par[x],par)
end

def same(x,y,par)
	return find(x,par) == find(y,par)
end

def union(x,y,par)
	x = find(x,par)
	y = find(y,par)
	return if x == y
	par[x] = y
end

edges = [[20,0,1],[3,1,2],[100,2,0]]
edges.sort!

def kruskal(edges, par)
  res = 0
  edges.each do |edge|
  	if !same(edge[1],edge[2],par)
  		res += edge[0]
  		union(edge[1],edge[2],par)
  	end
  end
  return res
end

p kruskal(edges,par)